### PR TITLE
Model dev

### PR DESCRIPTION
**Pull request by liav2002**
**Description:**

This branch introduces the development and implementation of the `Class Diagram ` I provides in the previous branch.

Key updates include:

1. **Class Diagram Design and Implementation:**
    -  Designed the class diagram to outline the project's architecture, modularity and scalability supported.
    -  Implemented the core classes (`Trainer`, `SiameseNetwork`, `Logger`, `PairDataset`) as per the designed architecture.
2. **Trainer Class:**
    -  Manages the training and validation loops, checkpointing, and early stopping.
    -  Includes dynamic data preparation, logging, and metric tracking (e.g., loss, accuracy, precision, recall, F1).
3. **SiameseNetwork Class:**
    -  Implements the Siamese Neural Network architecture based on the class diagram.
    -  Dynamically intialized using configurations defined in `config.yaml`.
    -  Supports weight intialization as per the referenced research paper.
4. **Logger Class:**
    -  Centralized logging for console, file, and MLFlow tracking.
    -  Integrated with all components to ensure clear and consistent reporting.
5. **Config-Driven Approach:**
    -  Centralized configuration in `config.yaml` for parameters like `learning_rate`, `augmentaions`, `logging_settings`, `early_stopping`, etc.
    -  Facilitates easy tuning and reproducibility.
6. **Data Handling:**
    -  Ensured training and validation datasets consist of unique individuals, avoiding data leakage.
    -  Handled edge cases to maximize the use of available pairs without compromising validation integrity.
7. **Logging and Metric Improvements:** 
    -  Enhanced logging to track key metrics (loss, accuracy, F1, precision, recall) during training and validation.
    -  Added MLFlow integration for experiment tracking and visualization.
8. **Testing and Debugging:**      
    -  Verified the architecture's functionality through iterative testing.
    -  Addressed early stopping behavior and refined training dynamics for better generalization.

### CODE REVIEW

**Review by DanielGoman — CHANGES_REQUESTED** [review]
Good stuff, clean and consice 🐱‍🚀

File: data/pairs_dataset.py
[the remark this answers is not in the document]
Comment (liav2002) [reply data/pairs_dataset.py:1]: I put the lfw2 folder on gitignore and removed that files from the repo.

The text files describe the split, and pdf file describe the format of those text files - I keep in the repo from the reason that they are not accessible from the internet.

File: .gitignore
[the remark this answers is not in the document]
Comment (liav2002) [reply .gitignore:1]: I move the components implementation to src/ folder execpt the unitests, I think they should be on the root folder.

File: config/config.yaml [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+      mean: 0.0
+      std: 0.2
+    fc_biases:
+      mean: 0.5
+      std: 0.01
+
+training:
+  batch_size: 128
+  num_epochs: 200
+  learning_rate: 0.00001
```
[the remark this answers is not in the document]
Comment (liav2002) [reply config/config.yaml]: Done

File: scripts/test.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+from config.consts import CONFIG_PATH
+from trainer.trainer import Trainer
+
+
+def test_model(checkpoint_path: str):
+    """
+    Test the Siamese Network on the test dataset.
+
+    Args:
+        config_path (str): Path to the YAML configuration file.
```
[the remark this answers is not in the document]
Comment (liav2002) [reply scripts/test.py]: Fixed.

File: scripts/train.py
[the remark this answers is not in the document]
Comment (liav2002) [reply scripts/train.py:1]: Done (I fix some typos errors and pep8 errors)

File: trainer/trainer.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+                transform_list.append(transforms.RandomHorizontalFlip(p=transform_config['probability']))
+            elif transform_type == "RandomRotation":
+                transform_list.append(transforms.RandomRotation(degrees=transform_config['degrees']))
+            elif transform_type == "RandomAdjustSharpness":
+                transform_list.append(transforms.RandomAdjustSharpness(
+                    sharpness_factor=transform_config['sharpness_factor'],
+                    p=transform_config['probability']
+                ))
+            elif transform_type == "ToTensor":
+                transform_list.append(transforms.ToTensor())
```
[the remark this answers is not in the document]
Comment (liav2002) [reply trainer/trainer.py]: Maybe?

I dont think so..

[the remark this answers is not in the document]
Comment (liav2002) [reply trainer/trainer.py]: Now it's a default transformation (don't depend on the config file)

File: trainer/trainer.py
[the remark this answers is not in the document]
Comment (liav2002) [reply trainer/trainer.py:1]: Fixed.

File: trainer/trainer.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+            return self._regularized_loss
+        else:
+            raise ValueError(f"Unsupported loss function: {loss_function_name}")
+
+    def _regularized_loss(self, outputs, labels):
+        cross_entropy = labels * torch.log(outputs) + (1 - labels) * torch.log(1 - outputs)
+        l2_reg = torch.tensor(0., requires_grad=True).to(self.device)
+        for param in self.model.parameters():
+            l2_reg = l2_reg + torch.norm(param, 2)
+        return -cross_entropy.mean() + self.config['training']['weight_decay'] * l2_reg
```
[the remark this answers is not in the document]
Comment (liav2002) [reply trainer/trainer.py]: The point has been made clear.

File: trainer/trainer.py [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+        if optimizer_name == "SGD":
+            self.logger.log_message(
+                f"SGD optimizer defined with lr={learning_rate}, momentum={momentum}, weight_decay={weight_decay}")
+            return optim.SGD(self.model.parameters(), lr=learning_rate, momentum=momentum, weight_decay=weight_decay)
+        elif optimizer_name == "Adam":
+            self.logger.log_message(
+                f"Adam optimizer defined with lr={learning_rate}, weight_decay={weight_decay}")
+            return optim.Adam(self.model.parameters(), lr=learning_rate, weight_decay=weight_decay)
+        else:
+            raise ValueError(f"Unsupported optimizer: {optimizer_name}")
```
[the remark this answers is not in the document]
Comment (liav2002) [reply trainer/trainer.py]: Fixed

File: model/siamese_network.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
```
[the remark this answers is not in the document]
Comment (liav2002) [reply model/siamese_network.py]: Removed

File: unitests/test_siamese_network.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    def test_forward_pass(self):
+        """Test the forward pass of the Siamese Network."""
+        output = self.model(self.x1, self.x2)
+        self.assertEqual(output.shape, (2, 1), "Output shape should be [B, 1]")
+        self.assertTrue(torch.all(output >= 0) and torch.all(output <= 1), "Output values should be between 0 and 1")
+
+    def test_shared_weights(self):
+        """Test if the twin networks share weights."""
+        cnn_params = list(self.model.cnn.parameters())
+        self.assertTrue(all(torch.equal(cnn_params[i], cnn_params[i]) for i in range(len(cnn_params))),
```
[the remark this answers is not in the document]
Comment (liav2002) [reply unitests/test_siamese_network.py]: It's a mistake

File: unitests/test_pair_dataset.py
[the remark this answers is not in the document]
Comment (liav2002) [reply unitests/test_pair_dataset.py:1]: Agree with that, noted.

**Comment by liav2002** [comment]
I have modified most of the code to suit your requirements, feel free to take a look.

File: .gitignore
[the remark this answers is not in the document]
Comment (DanielGoman) [reply .gitignore:1]: Alrighty.

Is there a reason you deleted your unittests? Generally you should always keep them, even if your code passes.

You never know when you're gonna make a change which suddenly creates some bug. Tests help you to automatically detect those stuff

Comment (DanielGoman) [inline .gitignore:1]: Well my Chrome almost crashed from the 13k changed files.

As a workaround idea, I think the better approach for this would be to create a separate branch from the source branch which contains those files (in this case - master), and dedicate that branch only to deleting them.

Merge to master, and then pull in your current branch.

Let me know fi you wanna discuss this / if this wasn't clear .-.

File: output/logs/log_2025-01-22_17-34-46.log
Comment (DanielGoman) [inline output/logs/log_2025-01-22_17-34-46.log:1]: outputs should be in `.gitignore`

File: .gitignore [hunk 8]
```diff
@@ -1,13 +1,10 @@
...
-data/__pycache__/
-notebooks/.ipynb_checkpoints/
-unitests/.pytest_cache/
-unitests/__pycache__/
-utils/__pycache__/
 config/__pycache__/
-model/__pycache__/
-scripts/__pycache__/
-trainer/__pycache__/
+data/__pycache__/
```
Comment (DanielGoman) [inline .gitignore:7]: `__pycache__` is already covered above

File: data/pairs_dataset.py
[the remark this answers is not in the document]
Comment (DanielGoman) [reply data/pairs_dataset.py:1]: They're available in our drive :)

File: config/consts.py [hunk 9]
```diff
@@ -1 +1,2 @@
-CONFIG_PATH = "./config/config.yaml"
\ No newline at end of file
+CONFIG_PATH = "./config/config.yaml"
+CHECKPOINT_PATH = "./output/checkpoints/best_model.pth"
```
Comment (DanielGoman) [inline config/consts.py:2]: Missing newline

File: src/trainer/trainer.py [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+
+        This function loads the image pairs from the specified text file, splits the pairs into training
+        and validation sets, applies the appropriate transformations, and initializes the data loaders
+        for both sets.
+
+        Args:
+            None
+
+        Returns:
+            None
```
Comment (DanielGoman) [inline src/trainer/trainer.py:107]: If there's no arguments then you can just skip this section in my opinion.

Same for return values

File: utils/logger.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+def get_logger(config: dict = None) -> Logger:
+    """
+    Retrieve the global Logger instance. If it does not exist, create it.
+
+    Args:
+        config (dict): Configuration dictionary. Required on first initialization.
+
+    Returns:
+        Logger: The global Logger instance.
```
[the remark this answers is not in the document]
Comment (DanielGoman) [reply utils/logger.py]: Didn't address this

File: src/trainer/trainer.py [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+
+        # Ensure all numeric parameters are correctly typed
+        if 'weight_decay' in optimizer_params:
+            optimizer_params['weight_decay'] = float(optimizer_params['weight_decay'])
+        if 'betas' in optimizer_params:
+            optimizer_params['betas'] = tuple(optimizer_params['betas'])
+        if 'lr' in optimizer_params:
+            optimizer_params['lr'] = float(optimizer_params['lr'])
+        if 'momentum' in optimizer_params:
+            optimizer_params['momentum'] = float(optimizer_params['momentum'])
```
Comment (DanielGoman) [inline src/trainer/trainer.py:158]: Sounds like something that maybe the proper place for it would be in the `config_loader.py` as it's responsible for proper loading of the config file

File: src/trainer/trainer.py [hunk 13]
```diff
@@ -0,0 +1,397 @@
+import os
+import torch
+import numpy as np
+import torch.nn as nn
+import torch.optim as optim
+from torch.optim import Optimizer
```
Comment (DanielGoman) [inline src/trainer/trainer.py:6]: Looks to me like one line to many

File: src/trainer/trainer.py [hunk 14]
```diff
@@ -0,0 +1,397 @@
...
+from torch.utils.data import DataLoader
+import torchvision.transforms as transforms
+from sklearn.metrics import precision_score, recall_score, f1_score
+from sklearn.metrics import roc_curve, roc_auc_score
+from typing import Union, Callable
```
Comment (DanielGoman) [inline src/trainer/trainer.py:11]: Some unused imports here

File: src/trainer/trainer.py [hunk 15]
```diff
@@ -0,0 +1,397 @@
+import os
+import torch
+import numpy as np
+import torch.nn as nn
+import torch.optim as optim
+from torch.optim import Optimizer
+from torch.utils.data import DataLoader
+import torchvision.transforms as transforms
+from sklearn.metrics import precision_score, recall_score, f1_score
```
Comment (DanielGoman) [inline src/trainer/trainer.py:9]: imports order. First `import`, then `from ... `

File: src/trainer/trainer.py [hunk 16]
```diff
@@ -0,0 +1,397 @@
+import os
+import torch
```
Comment (DanielGoman) [inline src/trainer/trainer.py:2]:
```suggestion
import os

import torch
```

File: model/siamese_network.py [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+        input_size = config['model']['input_size']
+        self.cnn = nn.Sequential(
+            nn.Conv2d(
+                in_channels=input_size[0],
+                out_channels=cnn_config['conv1']['out_channels'],
+                kernel_size=cnn_config['conv1']['kernel_size'],
+                stride=cnn_config['conv1']['stride']
+            ),
+            nn.ReLU(),
+            nn.MaxPool2d(kernel_size=cnn_config['conv1']['pool_size']),
```
[the remark this answers is not in the document]
Comment (DanielGoman) [reply model/siamese_network.py]: What do you think?

File: model/siamese_network.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+        self.apply(self._initialize_weights)
+
+    def _initialize_weights(self, module):
+        """
+        Initialize weights and biases for the network using configuration.
+
+        Args:
+            module (nn.Module): The layer to initialize.
+        """
+        if isinstance(module, nn.Conv2d):
```
[the remark this answers is not in the document]
Comment (DanielGoman) [reply model/siamese_network.py]: What do you think?

File: src/trainer/trainer.py [hunk 19]
```diff
@@ -0,0 +1,397 @@
...
+
+        Returns:
+            None
+        """
+        full_df = load_pairs_from_txt_file(
+            self.config['data']['train_pairs_path'],
+            self.config['data']['lfw_data_path']
+        )
+
+        train_df, val_df = split_pairs_by_connected_components(full_df, val_split=self.config["validation"]["val_split"],
```
Comment (DanielGoman) [inline src/trainer/trainer.py:114]: line too long

File: src/trainer/trainer.py [hunk 20]
```diff
@@ -0,0 +1,397 @@
...
+        if loss_function_name == "BinaryCrossEntropy":
+            self.logger.log_message(f"nn.BCELoss function defined as loss function.")
+            return nn.BCELoss()
+        elif loss_function_name == "RegularizedCrossEntropy":
+            self.logger.log_message(f"_regularized_loss function defined as loss function.")
+            return self._regularized_loss
+        else:
+            raise ValueError(f"Unsupported loss function: {loss_function_name}")
+
+    def _regularized_loss(self, outputs: torch.Tensor, labels: torch.Tensor)  -> torch.Tensor:
```
Comment (DanielGoman) [inline src/trainer/trainer.py:196]:
```suggestion
    def _regularized_loss(self, outputs: torch.Tensor, labels: torch.Tensor) -> torch.Tensor:
```

File: src/trainer/trainer.py [hunk 21]
```diff
@@ -0,0 +1,397 @@
...
+                all_predictions.extend(predictions.cpu().numpy())
+
+        val_loss /= len(self.val_loader)
+        accuracy = (np.array(all_predictions) == np.array(all_labels)).mean()
+        precision = precision_score(all_labels, all_predictions, zero_division=1)
+        recall = recall_score(all_labels, all_predictions, zero_division=1)
+        f1 = f1_score(all_labels, all_predictions, zero_division=1)
+
+        self.logger.log_message(
+            f"Validation: Loss = {val_loss:.4f}, Accuracy = {accuracy:.4f}, Precision = {precision:.4f}, Recall = {recall:.4f}, F1 = {f1:.4f}")
```
Comment (DanielGoman) [inline src/trainer/trainer.py:361]: line too long

**Review by DanielGoman — CHANGES_REQUESTED** [review]
Some more small comments 🐱‍👤
